src/main/java/fr/xephi/authme/listener/PlayerListener.java

```diff
@@ -103,103 +98,104 @@
     @Inject
     private BungeeSender bungeeSender;
 
-    private boolean isAsyncPlayerPreLoginEventCalled = false;
-    private Set<String> unresolvedPlayerHostname = new HashSet<>();
-
-    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOWEST)
-    public void onPlayerCommandPreprocess(PlayerCommandPreprocessEvent event) {
-        String cmd = event.getMessage().split(" ")[0].toLowerCase();
-        if (settings.getProperty(HooksSettings.USE_ESSENTIALS_MOTD) && "/motd".equals(cmd)) {
+    // Lowest priority to apply fast protection checks
+    @EventHandler(priority = EventPriority.LOWEST)
+    public void onAsyncPlayerPreLoginEventLowest(AsyncPlayerPreLoginEvent event) {
+        if (event.getLoginResult() != AsyncPlayerPreLoginEvent.Result.ALLOWED) {
             return;
         }
-        if (settings.getProperty(RestrictionSettings.ALLOW_COMMANDS).contains(cmd)) {
-            return;
-        }
-        final Player player = event.getPlayer();
-        if (!quickCommandsProtectionManager.isAllowed(player.getName())) {
-            event.setCancelled(true);
-            player.kickPlayer(messages.retrieveSingle(player, MessageKey.QUICK_COMMAND_PROTECTION_KICK));
+        final String name = event.getName();
+
+        // NOTE: getAddress() sometimes returning null, we don't want to handle this race condition
```

Comment:
I really like this comment! This method would benefit of some tests, though :)  (only the try-catch stuff below is covered)

Reply:
Hmm, could you please help me creating this test?

Reply:
Of course! If you look at PlayerListener#shouldPerformAllJoinVerificationsSuccessfullyPreLoginLowest you can basically just copy it BUT create the `preLoginEvent` to have `null` as IP address. Also because a message is taken from the messages service, mock it to return something, i.e. under "given" add `given(messages.retrieveSingle(name, MessageKey.KICK_UNRESOLVED_HOSTNAME)).willReturn("Unresolved hostname");`

Afterwards under "then" you can `verify(event).disallow(AsyncPlayerPreLoginEvent.Result.KICK_OTHER, "Unresolved hostname")` because that's exactly what you expect to have been called on the event. I would also add a `verifyNoInteractions(onJoinVerifier)` so the test makes sure we don't do anything with it in this case